flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java

```diff
@@ -57,39 +70,313 @@
+     *       required in case of DeltaSink.
+     * </ol>
+     */
     List<DeltaCommittable> prepareCommit(boolean flush) throws IOException {
-        return Collections.emptyList();
+        if (deltaInProgressPart != null) {
+            if (rollingPolicy.shouldRollOnCheckpoint(deltaInProgressPart.getInProgressPart())
+                || flush) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug(
+                        "Closing in-progress part file for bucket id={} on checkpoint.",
+                        bucketId);
+                }
+
+                closePartFile();
+            } else {
+                throw new RuntimeException(
+                    "Unexpected behaviour. Bulk format writers should always roll part files " +
+                        "on checkpoint. To resolve this issue verify behaviour of your" +
+                        " rolling policy.");
+            }
+        }
+
+        List<DeltaCommittable> committables = new ArrayList<>();
+        pendingFiles.forEach(pendingFile -> committables.add(
+            new DeltaCommittable(pendingFile)));
+        pendingFiles.clear();
+
+        return committables;
     }
 
+    /**
+     * This method is responsible for snapshotting state of the bucket writer. The writer's
+     * state snapshot can be further used to recover from failure or from manual Flink's app
+     * snapshot.
+     * <p>
+     * Since the writer is supposed to always roll part files on checkpoint then there is not
+     * much state to snapshot and recover from except bucket metadata (id and path) and also
+     * unique identifier for the application that the writer is part of.
+     *
+     * @return snapshot of the current bucket writer's state
+     */
     DeltaWriterBucketState snapshotState() {
-        return new DeltaWriterBucketState();
+        return new DeltaWriterBucketState(
+            bucketId,
+            bucketPath
+        );
+    }
+
+    /**
+     * Method responsible for "closing" previous in-progress file and "opening" new one to be
+     * written to.
+     *
+     * @param currentTime current processing time
+     * @return new in progress part instance representing part file that the writer will start
+     * write data to
+     * @throws IOException Thrown if the writer cannot be opened, or if the output stream throws an
+     *                     exception.
+     * @implNote This method behaves in the similar way as
+     * org.apache.flink.connector.file.sink.writer.FileWriterBucket#rollPartFile
+     * except that it uses custom implementation to represent the in-progress part file.
+     * See {@link DeltaInProgressPart} for details.
+     */
+    private DeltaInProgressPart<IN> rollPartFile(long currentTime) throws IOException {
+        closePartFile();
+
+        final Path partFilePath = assembleNewPartPath();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(
```

Comment:
nit: is there no way to move these `LOG.isDebugEnabled` checks to some common `Logging` class?
i.e. most codes just say `logDebug(...)` and then the Logging class performs the `LOG.isDebugEnabled` check for us.

Reply:
Great point! Then this as it is is fine :)